kaldb/src/main/java/com/slack/kaldb/server/OpenSearchBulkIngestApi.java

```diff
@@ -0,0 +1,230 @@
+      // We think most indexing requests will be against 1 index
+      if (docs.keySet().size() > 1) {
+        BulkIngestResponse response =
+            new BulkIngestResponse(0, 0, "request must contain only 1 unique index");
+        return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+      }
+
+      for (Map.Entry<String, List<Trace.Span>> indexDocs : docs.entrySet()) {
+        final String index = indexDocs.getKey();
+        if (!rateLimiterPredicate.test(index, indexDocs.getValue())) {
+          BulkIngestResponse response = new BulkIngestResponse(0, 0, "rate limit exceeded");
+          return HttpResponse.ofJson(TOO_MANY_REQUESTS, response);
+        }
+      }
+      BulkIngestResponse response = produceDocuments(docs);
+      return HttpResponse.ofJson(response);
+    } catch (Exception e) {
+      LOG.error("Request failed ", e);
+      BulkIngestResponse response = new BulkIngestResponse(0, 0, e.getMessage());
+      return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+    }
+  }
+
+  public BulkIngestResponse produceDocuments(Map<String, List<Trace.Span>> indexDocs) {
+    int totalDocs = indexDocs.values().stream().mapToInt(List::size).sum();
+
+    String transactionId = UUID.randomUUID().toString();
+    // we cannot create a generic pool of producers because the kafka API expects the transaction ID
+    // to be a property while creating the producer object.
+    try (KafkaProducer<String, byte[]> kafkaProducer =
+        createKafkaTransactionProducer(transactionId)) {
+      for (Map.Entry<String, List<Trace.Span>> indexDoc : indexDocs.entrySet()) {
+        String index = indexDoc.getKey();
+        // call once per batch and use the same partition for better batching
+        int partition = getPartition(index);
+
+        // since there isn't a dataset provisioned for this service/index we will not index this set
+        // of docs
+        if (partition < 0) {
+          LOG.warn("index=" + index + " does not have a provisioned dataset associated with it");
+          continue;
+        }
+        try {
+          kafkaProducer.initTransactions();
+          kafkaProducer.beginTransaction();
+          for (Trace.Span doc : indexDoc.getValue()) {
+
+            ProducerRecord<String, byte[]> producerRecord =
+                new ProducerRecord<>(
+                    preprocessorConfig.getDownstreamTopic(), partition, index, doc.toByteArray());
+            kafkaProducer.send(producerRecord);
+          }
+          kafkaProducer.commitTransaction();
+        } catch (Exception e) {
+          try {
+            kafkaProducer.abortTransaction();
+          } catch (ProducerFencedException err) {
+            LOG.error("Could not abort transaction", err);
+          }
+          return new BulkIngestResponse(0, totalDocs, e.getMessage());
+        }
+      }
+    }
+
+    return new BulkIngestResponse(totalDocs, 0, "");
+  }
+
+  private int getPartition(String index) {
```

Comment:
I think we do this because it's throughput based? I actually don't understand the full value on how logs can belong to multiple service datasets ( in a practical sense ) which is why we use this technique.

Reply:
This is fine per discussion - this is because a given service name can match multiple, depending on glob/regex matching.